### PR TITLE
feat: toggle follow fiori3

### PR DESCRIPTION
**Pull request by JKMarkowski**
## Related Issue
Closes SAP/fundamental-styles#456
Closes https://github.com/SAP/fundamental-styles/issues/535

## Description
There are is new toggle, which follows fiori3 specs

## Breaking changes
* Completely new structure of toggle in every mode. Added classes: 
`fd-toggle__text` - To put some text (max 3 chars) inside toggle.
`fd-toggle__wrapper`, `fd-toggle__track`.
* Removed size classes (--xs, --s, --m, --l, --xl)

### After:
image
image
image
image

### CODE REVIEW

File: src/toggle.scss [hunk 1]
```diff
@@ -11,35 +11,89 @@ $block: #{$fd-namespace}-toggle;
...
+
+    font-size: var(--sapFontSmallSize);
+    line-height: 1.375rem;
+    display: block;
+    width: $fd-toggle-label-width;
+    text-transform: uppercase;
+    text-align: center;
+  }
+
+  &__switch {
```
[the remark this answers is not in the document]
Comment (JKMarkowski) [reply src/toggle.scss]: shouldn't it be text with that margin?

[the remark this answers is not in the document]
Comment (JKMarkowski) [reply src/toggle.scss]: Done

File: src/toggle.scss [hunk 2]
```diff
@@ -11,35 +11,89 @@ $block: #{$fd-namespace}-toggle;
...
-  $fd-toggle-switch-background-color--disabled: hsla(240, 2%, 93%, 1) !default;
-  $fd-toggle-switch-border-color--disabled: hsla(212, 8%, 57%, 0.4) !default;
-  $fd-toggle-switch-on-background-color--disabled: #9ebbd8 !default;
-  $fd-toggle-switch-hover-background-color: #ebf5fe;
+  $fd-toggle-active-border-color: var(--sapButton_Selected_BorderColor);
+  $fd-toggle-active-background-color: var(--sapButton_Selected_Background);
+  $fd-toggle-inactive-border-color: var(--sapContent_ForegroundBorderColor);
+  $fd-toggle-inactive-switch-background: var(--sapButton_Background);
+
+  $fd-toggle-hover-border-color: var(--sapButton_Hover_BorderColor);
```
[the remark this answers is not in the document]
Comment (JKMarkowski) [reply src/toggle.scss]: It's used several times, that's why I made variable from this

File: src/toggle.scss [hunk 3]
```diff
@@ -11,35 +11,89 @@ $block: #{$fd-namespace}-toggle;
...
 .#{$block} {
 
-  $fd-toggle-sizes: ("default": fd-space(13) fd-space(5.5) 20px -5px -5px,
-  "xs": fd-space(9) fd-space(5) 14px -4.5px 15px,
-  "s": fd-space(10) fd-space(6) 17px -5px 15px,
-  "compact": fd-space(10) fd-space(6.5) 16px -6px 13px,
-  "l": fd-space(14) fd-space(8) 22px -6px 23px,
-) !default;
-  $_defaults: map-get($fd-toggle-sizes, "default");
-  $fd-toggle-width: nth($_defaults, 1);
```
[the remark this answers is not in the document]
Comment (JKMarkowski) [reply src/toggle.scss]: Done

File: src/form-label.scss [hunk 4]
```diff
@@ -84,5 +84,8 @@ $block: #{$fd-namespace}-form-label;
     overflow: initial;
     margin-top: 5px;
     margin-bottom: 5px;
+    display: flex;
```
Comment (stefanoScalzo) [inline src/form-label.scss]: why is this needed?

File: docs/pages/components/toggle.md [hunk 5]
```diff
@@ -14,189 +14,171 @@ The toggle/switch is meant to resemble a physical switch and allow a user to tur
...
-To clarify the meaning of the toggle, use a label above it (ex. Live).
+## Toggle with Labels on Sides
+This is basic usage of toggle
 
 {% capture default %}
 <div class="fd-form-group">
     <div class="fd-form-item">
-        <label class="fd-form-label" for="y21YO391">Live:</label>
+        <label class="fd-form-label">Toggle with label</label>
         <label class="fd-form-label fd-form-label--toggle">
```
Comment (InnaAtanasova) [inline docs/pages/components/toggle.md]: This will make the toggle component dependent on `fd-form-label`. Can you create a new class and if you depend a lot on `fd-form-label` maybe create a mixin and include it here.

File: docs/pages/components/toggle.md [hunk 6]
```diff
@@ -14,189 +14,171 @@ The toggle/switch is meant to resemble a physical switch and allow a user to tur
...
             <span class="fd-toggle">
-                <input class="fd-toggle__input" type="checkbox" name="" value="" id="y21YO391">
-                <span class="fd-toggle__switch" role="presentation"></span>
+                <input class="fd-toggle__input" type="checkbox" name="" value="" id="y21YO3911">
+                <div class="fd-toggle__wrapper">
+                    <div class="fd-toggle__inner">
+                        <span class="fd-toggle__label fd-toggle__label--on">on</span>
```
Comment (InnaAtanasova) [inline docs/pages/components/toggle.md]: `fd-switch__text`

File: docs/pages/components/toggle.md [hunk 7]
```diff
@@ -14,189 +14,171 @@ The toggle/switch is meant to resemble a physical switch and allow a user to tur
...
+                        <span class="fd-toggle__switch" role="presentation"></span>
```
Comment (InnaAtanasova) [inline docs/pages/components/toggle.md]: `fd-switch__handle`

File: docs/pages/components/toggle.md [hunk 8]
```diff
@@ -14,189 +14,171 @@ The toggle/switch is meant to resemble a physical switch and allow a user to tur
...
     <div class="fd-form-item">
-        <label class="fd-form-label" for="y21YO391">Live:</label>
+        <label class="fd-form-label">Toggle with label</label>
         <label class="fd-form-label fd-form-label--toggle">
             <span class="fd-toggle">
-                <input class="fd-toggle__input" type="checkbox" name="" value="" id="y21YO391">
-                <span class="fd-toggle__switch" role="presentation"></span>
+                <input class="fd-toggle__input" type="checkbox" name="" value="" id="y21YO3911">
+                <div class="fd-toggle__wrapper">
+                    <div class="fd-toggle__inner">
```
Comment (InnaAtanasova) [inline docs/pages/components/toggle.md]: `fd-switch__track`

File: src/toggle.scss [hunk 9]
```diff
@@ -11,35 +11,114 @@ $block: #{$fd-namespace}-toggle;
...
+      margin: 0 0.125rem 0 0;
+    }
+  }
+
+  &__switch {
+    border: $fd-toggle-border-width solid $fd-toggle-inactive-border-color;
+    background-color: $fd-toggle-inactive-switch-background;
+    width: $fd-toggle-switch-size;
+    height: $fd-toggle-switch-size;
+    border-radius: $fd-toggle-switch-border-radius;
```
Comment (InnaAtanasova) [inline src/toggle.scss]: missing `background-clip: padding-box;`

File: src/toggle.scss [hunk 10]
```diff
@@ -11,35 +11,114 @@ $block: #{$fd-namespace}-toggle;
...
+  $fd-toggle-border-width: var(--sapButton_BorderWidth);
+
+  $fd-toggle-label-width: 1.75rem;
+
+  $fd-toggle-transition: all 0.1s;
+
+  $fd-toggle-container-padding-y: 0.375rem;
+  $fd-toggle-size-x: 3.75rem;
+  $fd-toggle-size-y: 1.25rem;
+  $fd-toggle-switch-size: 1.875rem;
```
Comment (InnaAtanasova) [inline src/toggle.scss]: Isn't it 2rem?

Comment (JKMarkowski) [reply src/toggle.scss]: It is, but actually it's counted with (2* 0.0625rem) border, so it gives 2rem.

File: docs/pages/components/toggle.md (hunk 6 above)
Comment (JKMarkowski) [reply docs/pages/components/toggle.md]: I wanted to use `fd-toggle__label` not only for text, it's also for icons.

File: docs/pages/components/toggle.md (hunk 7 above)
Comment (JKMarkowski) [reply docs/pages/components/toggle.md]: You mean `fd-toggle__handle`?

File: docs/pages/components/toggle.md (hunk 6 above)
Comment (JKMarkowski) [reply docs/pages/components/toggle.md]: I will use `__text`, because `__label` is reserved for labels

**Comment by JKMarkowski** [comment]
All of the comments are solved now. It's ready to second review.

File: docs/pages/components/toggle.md [hunk 11]
```diff
@@ -14,189 +14,171 @@ The toggle/switch is meant to resemble a physical switch and allow a user to tur
...
 
 {% capture default %}
 <div class="fd-form-group">
     <div class="fd-form-item">
-        <label class="fd-form-label" for="y21YO391">Live:</label>
-        <label class="fd-form-label fd-form-label--toggle">
-            <span class="fd-toggle">
-                <input class="fd-toggle__input" type="checkbox" name="" value="" id="y21YO391">
-                <span class="fd-toggle__switch" role="presentation"></span>
+        <label class="fd-form-label">Toggle with label</label>
```
Comment (stefanoScalzo) [inline docs/pages/components/toggle.md]: shouldn't you only use the switch label?

Comment (JKMarkowski) [reply docs/pages/components/toggle.md]: Labels are not stricly connected to `switch`.

**Comment by InnaAtanasova** [comment]
1) The elements are not focusable and the keyboard support is missing
2) There's bleeding of styles - check the un-normalize in the Playground
Screen Shot 2020-01-14 at 3 21 17 PM
Screen Shot 2020-01-14 at 3 13 49 PM

**Comment by InnaAtanasova** [comment]
Screen Shot 2020-01-15 at 1 39 55 PM
Screen Shot 2020-01-15 at 1 38 54 PM

The tests are failing because the "handle" moved a bit (prob because of the resets). I think it's related to the `left: calc(1.8125rem - 100%);` for `fd-switch__track`.

**Comment by bcullman** [comment]
can we see a before and after screen shots on the PR description?

**Comment by JKMarkowski** [comment]
@bcullman I noticed your comments too late, after merging this commit. But it has been tested and compared to fiori3 specs and everything works correct.

**Comment by bcullman** [comment]
@thanks @JKMarkowski looks great! Thats been wrong for a *long* time. Glad thats been fixed.

I did note one other oddity on the alert - when in RTL mode, the close button is off by pixel from the edge. See

image

vs rtl

image
